src/util.c

```diff
@@ -871,3 +871,33 @@ netplan_state_iterator_has_next(const NetplanStateIterator* iter)
         return FALSE;
     return _iter->next != NULL;
 }
+
+/*
+ * Returns true if a route already exists in the netdef routes list.
+ *
+ * We consider a route a duplicate if it is in the same table, has the same metric,
+ * src, to and via values.
+ *
+ * XXX: in the future we could add a route "key" to a hash set so this verification could
+ * be done faster.
+ */
+gboolean
+is_route_present(const NetplanNetDefinition* netdef, const NetplanIPRoute* route)
+{
+
+    const GArray* routes = netdef->routes;
+
+    for (int i = 0; i < routes->len; i++) {
+        const NetplanIPRoute* entry = g_array_index(routes, NetplanIPRoute*, i);
+        if (
+                entry->table == route->table &&
+                entry->metric == route->metric &&
+                g_strcmp0(entry->from, route->from) == 0 &&
+                g_strcmp0(entry->to, route->to) == 0 &&
+                g_strcmp0(entry->via, route->via) == 0
+           )
```

Comment:
Beware that `->to` could be `default` or `0.0.0.0/0` / `::/0` which is logically the same and should fail this check.
We might want to compare `->family`, too. (IPv4 vs IPv6)

Furthermore, there can be routes which don't have the "via" field set... (e.g. when using "scope: link"), so I feel like we need to think a bit more about how to properly compare routes.

Reply:
Ack. I tried to handle these cases. For fields that might be NULL, g_strcmp0() will handle it.